### PR TITLE
Refactors anti-magic into a component

### PR DESCRIPTION
**Pull request by XDTM**
:cl: XDTM
balance: Wizard spells and items can now be resisted/ignored with anti-magic items/clothing such as null rods!
balance: Revenant spells can now be resisted with "holy" items like null rods and bibles.
balance: Wizard hardsuits are now magic immune, but not holy.
balance: Immortality Talismans now grant both spell and holy immunity.
tweak: Inquisitor Hardsuits already granted spell and holy immunity, but now they do it properly instead of having a null rod embedded inside.
tweak: Holy Melons now grant holy immunity.
/:cl:

@vuonojenmustaturska here you go

Makes anti-magic a far less snowflake thing by using a component. Spells now cycle through the target's inventory looking for an item with that component, and if so it checks the relative resistance (currently there's either magic or holy), and returns the item in case you want to do something with it.

Small note: the bible_check for some reason was in a blood cult file while it's only used for ocular wardens. I moved it inside the bible file instead. I did not replace it since there was code for the bible bursting on fire and so on that i couldn't replicate by just using the component.

### CODE REVIEW

**Comment by KorPhaeron** [comment]
Great PR, thanks

**Comment by XDTM** [comment]
I'm probably missing some items that should either be anti-magic or holy, if someone points them out i'll add them

**Comment by JJRcop** [comment]
You could optimize this a little bit by creating another component that is given directly to the mob by the anti magic component, holding weakrefs to the anti-magic items that added it, then the new component has the anti magic check proc that just looks through those weakrefs to see if they are still on the mob and still have the anti magic component, so it doesn't need to check all the items on the mob.

This setup you have is simple, however, so the performamce benefit may not be worth the extra complexity.

**Comment by XDTM** [comment]
I'm leaning towards code simplicity since it only checks about ten items at most. If it turns out to be too heavy i'll try that, though.

**Comment by Robustin** [comment]
Only thing im confused about is holy melons, does that mean it behaves differently than holy water? If so, why?

**Comment by XDTM** [comment]
Holy water currently offers no protection against magic and has no relation to this system (yet). I added holy protection to holymelons, so if you're holding one you're immune to stuff like revenants and cult conversion. 

I kinda like the idea of someone hugging a watermelon for dear life while cultists around him are trying to tug it away from him

**Comment by Dorsisdwarf** [comment]
Do watermelons only work in hands? because hugging a melon to not die is fun, backpack melons for the entire crew  to be cultproof isnt

**Comment by JJRcop** [comment]
You could add a "strong" var to the component, if true it works if you have the item on you at all, otherwise if it's false, the item must be directly equipped.

**Comment by YPOQ** [comment]
Seems kind of lame to hard-counter things like ocular wardens and revenant abilities with an item mass-producible at round start.

**Comment by Kromgar** [comment]
So if i have an immortality talisman in my pocket im magic immune?

**Comment by Durkel** [comment]
Only in ss13 would you see a grown man running full speed down the hallways while lugging a eight pound water melon in order to ward off violent ghosts of the departed.

**Comment by ShadowDeath6** [comment]
can't wait for the sticky botany trait to make the melons nodrop, great pr

**Comment by JJRcop** [comment]
@Dorsisdwarf So this just checks the items that are equipped in one of your slots (or hands), otherwise it doesn't do any deep diving or anything for stuff in your backpack, that means you have to hold the holy water melon.

**Comment by avosirenfal** [comment]
I'm so in love with the holy melon.

**Comment by Tlaltecuhtli** [comment]
uhh why does chaplain need more anti specific antags?

**Comment by XDTM** [comment]
I'm also making it not chaplain-only, but null rod protecting against magic is pretty established

**Comment by Kromgar** [comment]
image

File: code/game/objects/items/holy_weapons.dm [hunk 1]
```diff
@@ -1,3 +1,32 @@
+/datum/component/anti_magic
```
Comment (ninjanomnom) [inline code/game/objects/items/holy_weapons.dm]: this needs to be in the component directory

File: code/game/objects/items/holy_weapons.dm [hunk 2]
```diff
@@ -1,3 +1,32 @@
...
+	var/list/obj/item/item_list = list()
+	if(isliving(src))
+		var/mob/living/L = src
+		item_list |= L.get_equipped_items(TRUE)
+	item_list |= held_items
+	for(var/obj/O in item_list)
+		GET_COMPONENT_FROM(anti_magic, /datum/component/anti_magic, O)
+		if(!anti_magic)
+			continue
+		if((magic && anti_magic.magic) || (holy && anti_magic.holy))
```
Comment (ninjanomnom) [inline code/game/objects/items/holy_weapons.dm]: do the logic inside the component, best bet is probably to check a signal return flag

**Review by ninjanomnom — DISMISSED** [review]

**Comment by XDTM** [comment]
I'm not sure how to set up a signal in a way that it bounces from the mob onto each item they're wearing/holding and checking if one returns TRUE, or if doing so is worth it, considering the anti_magic_check proc is gonna stay anyway as it needs to consider traits as well.

I did move the component (and relative mob procs) to the component folder, and now the component can be disabled, instead of having to remove it, in case there are items with conditional or temporary magic resistance.

**Review by Cyberboss — DISMISSED** [review]

File: code/datums/components/anti_magic.dm [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	item_list |= held_items
+	for(var/obj/O in item_list)
+		GET_COMPONENT_FROM(anti_magic, /datum/component/anti_magic, O)
+		if(!anti_magic)
+			continue
+		if(anti_magic.can_protect(magic, holy))
+			return O
+
+/mob/living/anti_magic_check(magic = TRUE, holy = FALSE)
+	if((magic && has_trait(TRAIT_ANTIMAGIC)) || (holy && has_trait(TRAIT_HOLY)))
```
[the remark this answers is not in the document]
Comment (XDTM) [reply code/datums/components/anti_magic.dm]: traits have better source tracking, this way, for example, holy water can make you holy and when it expires it won't remove holiness that comes from being an angel

**Pull request merged**